library/std/src/sync/mutex.rs

```diff
@@ -546,9 +596,158 @@ impl<T: ?Sized + fmt::Display> fmt::Display for MutexGuard<'_, T> {
 }
 
 pub fn guard_lock<'a, T: ?Sized>(guard: &MutexGuard<'a, T>) -> &'a sys::Mutex {
-    &guard.lock.inner
+    &guard.lock.state.inner
 }
 
 pub fn guard_poison<'a, T: ?Sized>(guard: &MutexGuard<'a, T>) -> &'a poison::Flag {
-    &guard.lock.poison
+    &guard.lock.state.poison
+}
+
+impl<'a, T: ?Sized> MutexGuard<'a, T> {
+    /// Makes a [`MappedMutexGuard`] for a component of the borrowed data, e.g.
+    /// an enum variant.
+    ///
+    /// The `Mutex` is already locked, so this cannot fail.
+    ///
+    /// This is an associated function that needs to be used as
+    /// `MutexGuard::map(...)`. A method would interfere with methods of the
+    /// same name on the contents of the `MutexGuard` used through `Deref`.
+    #[unstable(feature = "mapped_lock_guards", issue = "117108")]
+    pub fn map<U, F>(orig: Self, f: F) -> MappedMutexGuard<'a, U>
+    where
+        F: FnOnce(&mut T) -> &mut U,
+        U: ?Sized,
+    {
+        let mut orig = ManuallyDrop::new(orig);
+        let value = NonNull::from(f(&mut *orig));
```

Comment:
Doesn't this leak `orig` if `f` panics, leaving the lock permanently locked? I would also add a test for `mutex_lock.map(|_| panic!())` or something like that.

The same applies to the other implementations.

Reply:
Yeah that makes sense. I've refactored it so that the "old" guard is only wrapped in `ManuallyDrop` after the closure completes, so panicking in the closure has the same effect as panicking while holding the guard normally (unlocks the guard; poisons for `MutexGuard` and `RwLockWriteGuard`, doesn't poison for `RwLockReadGuard`). I've also added tests for panicking in the closures.